### PR TITLE
[Core] ClearSolutionStepData in node destructor (and unique VariablesList in ConnectivityPreserveModeler)

### PR DESCRIPTION
**Pull request by rubenzorrilla**
This PR modifies the `ConnectivityPreserveModeler` to have a unique variables list (the destination model part one points to the same one as the origin model part points to).

This required to add a new method to the `ModelPart` to get a pointer to the variables list.

### CODE REVIEW

**Review by RiccardoRossi — DISMISSED** [review]
+1 as spoken also with pooyan

**Comment by rubenzorrilla** [comment]
Wait... the `VariationalDistanceCalculationProcess` crashes.

**Review by ddiezrod — CHANGES_REQUESTED** [review]
Blocking as variational is failing. It seems redistance is computing fine (if you check nodal values at the end of the redistance they are correct) but then after the object is destroyed the nodal values from the main model part are not initialized.

**Comment by rubenzorrilla** [comment]
@ddiezrod I observed the same. I'll further investigate tomorrow.

**Comment by philbucher** [comment]
FYI @sunethwarna

**Comment by jcotela** [comment]
Related to #4979 and #4267 I guess... It would be great if we could decide on a consistent way of managing the pVariablesList when we have two ModelParts with the same nodes.

**Comment by philbucher** [comment]
What was the decision?

**Comment by RiccardoRossi** [comment]
the proposal of @pooyan-dadvand is that the ConnectivityPreserveModeler replaces the variable list of the modelpart being filled.

on my side it makes sense, nevertheless i think there are still cases in which this is not enough

In my opinion in the VariableUtils::AddDofs we should also check if multiple variable lists exist and ensure that dofs are added everywhere. 
This is important since we may have modelparts with nodes coming from different modelparts and generated differently than connectivity_preserve_modeler

also i have the feeling that the problem we have with variational_distance is related to the fact that in there we destroy the modelpart. @marandra i cannot find your issue related to problems in destruction. could u mention it here?

**Comment by jcotela** [comment]
@RiccardoRossi @pooyan-dadvand shouldn't the variable list be the same for all model parts that contain the same nodes? Not a copy, not a duplicate with the same variables, but the same object in memory. Otherwise we are just asking for problems as soon as we allow for the same node to belong to two modelparts with different variable lists, since each modelpart will make different assumptions about the memory layout of the node's data.

**Comment by RiccardoRossi** [comment]
what happens if you have a modelpart, for example used for vidualization,
with nodes coming from.different sources?

**Comment by jcotela** [comment]
@RiccardoRossi how do you know that you are printing the same result on all nodes? (and not just reading from a random spot in memory)

**Comment by RiccardoRossi** [comment]
you can do since each node holds a pointer to its own variable list...

this is a supported feature (if we want to remove this capability we need
to think about this carefully)

**Comment by jcotela** [comment]
Ok let me rephrase my objection: the variables list is a property of the node (it says how the node organizes its solution step data). For efficiency reasons, we allow nodes to share a common variables list, and only keep a pointer to this shared variables list on each node. So far so good.

The decision was made to store this shared variable list on the modelpart. This is somehow justified by an assumption that all nodes on the modelpart will share the variable list, which is true for simple cases.

Now you say that you want to mix nodes with different variable lists on the same modelpart. Technically you can make it work (as you say) but this means that the variable list on the ModelPart becomes meaningless, since you cannot rely on the modelpart's variable list to say anything at all about the data stored on its nodes. Why are you storing the variable list on the model part then?

This is not a problem of the technical details (although I suspect that your problems with DOF come from the code relying on the wrong assumption that the modelpart's variable list is good for all of its nodes). Instead it is a design issue, which is giving us a lot of trouble in the connectivity preserve modeler and the variational distance calculator.

In my opinion, either we enforce the design principle of "the variable list of the modelpart is the variable list of all its nodes" or we completely remove the variable list from the modelpart, since they become completely unrelated objects.

I am sorry for derailing the PR, I guess I have strong feelings about this...

**Comment by RiccardoRossi** [comment]
@jcotela i understand what you are telling, nevertheless the point is that until now it was a supported feature. We may take the conscious decision of breaking the support, nevertheless we should have a way of providing equivalent capabilities.

the use case which is problematic in this case is visualization modelparts where nodes are coming from a lot of different sources. We cannot just get rid of this silently...

**Comment by rubenzorrilla** [comment]
Aside of the discussion above, I think we found the segfault origin.

In the destructor of the `VariationalDistanceCalculationProcess` we destroy the auxiliary model part we have created to compute the distance. 
https://github.com/KratosMultiphysics/Kratos/blob/80d68b9a7e89d97438fd04726e5198084b7be43e/kratos/processes/variational_distance_calculation_process.h#L186-L192
Note that such auxiliary model part is built using the `ConnectivityPreserveModeler`. The changes in this PR imply that both the origin and auxiliary model parts variables list pointers point to the same variables list.

In the model part destructor we call the `ClearSolutionStepData` for each node of the auxiliary model part. As a consequence the origin variables list is also destroyed.
https://github.com/KratosMultiphysics/Kratos/blob/80d68b9a7e89d97438fd04726e5198084b7be43e/kratos/sources/model_part.cpp#L60-L67
My question is: **why do we need to call this? Shouldn't it be destroyed when the last node that points to it is destroyed?**.

**Comment by philbucher** [comment]
Looking at the code I don’t understand the problem, wasn’t the var list already now the same in both MPs?

**Comment by rubenzorrilla** [comment]
> Looking at the code I don’t understand the problem, wasn’t the var list already now the same in both MPs?

Now the var list is unique for both model parts. Thus, when we call the destructor of one of them we enter here
https://github.com/KratosMultiphysics/Kratos/blob/80d68b9a7e89d97438fd04726e5198084b7be43e/kratos/sources/model_part.cpp#L60-L67
Taking into account that the nodes are the same between both model parts (due to the fact that one of them has been built with the `ConnectivityPreserveModeler`) we are destroying the origin model part nodal variables list at the same time.

**Comment by philbucher** [comment]
I meant what is the motivation for this PR?
Looking at the code it sees to be the same behavior just using a different method

**Comment by ddiezrod** [comment]
@philbucher The difference is that before this MR, the new model part created a copy of the original model part variable list. The idea is that as the nodes have a pointer to the original model part variable list, then it makes sense that the new model part also points to the same object instead of having a copy.

As @rubenzorrilla points out, the problem is that when you destroy the redistance model part, it cleans the solution step data for both the redistance model part and the original model part and test fails. I guessed that it used to be two different variable lists this was not causing trouble before.

**Comment by rubenzorrilla** [comment]
> Aside of the discussion above, I think we found the segfault origin.
> 
> In the destructor of the `VariationalDistanceCalculationProcess` we destroy the auxiliary model part we have created to compute the distance.
> 
> https://github.com/KratosMultiphysics/Kratos/blob/80d68b9a7e89d97438fd04726e5198084b7be43e/kratos/processes/variational_distance_calculation_process.h#L186-L192
> 
> Note that such auxiliary model part is built using the `ConnectivityPreserveModeler`. The changes in this PR imply that both the origin and auxiliary model parts variables list pointers point to the same variables list.
> 
> In the model part destructor we call the `ClearSolutionStepData` for each node of the auxiliary model part. As a consequence the origin variables list is also destroyed.
> 
> https://github.com/KratosMultiphysics/Kratos/blob/80d68b9a7e89d97438fd04726e5198084b7be43e/kratos/sources/model_part.cpp#L60-L67
> 
> My question is: **why do we need to call this? Shouldn't it be destroyed when the last node that points to it is destroyed?**.

Definitively, this is not needed anymore. As discussed with @pooyan-dadvand and @RiccardoRossi, moving the `ClearSolutionStepData` to the node destructor solves the problem. Note that in this way the shared variable list will be destroyed only when the last node that points to it is destroyed.

**Review by RiccardoRossi — APPROVED** [review]
to me it is a GO.

**Review by RiccardoRossi — APPROVED** [review]
to me it is a GO.

**Comment by philbucher** [comment]
@ddiezrod but doing this renders having the varlist one the ModelPart useless I‘d say. What if I wanted other variables in one ModelPart?
I see that the current design is not good but I also don’t see an improvement with this PR
Why don’t we try to find a better solution for this rather than going from one hack to another?

**Comment by philbucher** [comment]
Either way I am not blocking, but I feel like this needs a better solution in the future to not create more problems than we already have ;)

**Comment by rubenzorrilla** [comment]
> @ddiezrod but doing this renders having the varlist one the ModelPart useless I‘d say. What if I wanted other variables in one ModelPart?
> I see that the current design is not good but I also don’t see an improvement with this PR
> Why don’t we try to find a better solution for this rather than going from one hack to another?

Actually I think that if you are using the `ConnectivityPreserveModeler` what you really want is that the variables in both the origin and destination model parts (nodes) are the same (mainly for volume coupling). With the changes in here we ensure that the variables list container is exactly the same rather than a copy.

I agree that this may introduce some overhead, as we add some variables that are not explicitly required. However, I can't figure out a way to make the `ConnectivityPreserveModeler` capable of distinguishing which variables are and are not needed...

**Comment by rubenzorrilla** [comment]
@ddiezrod now it should work. Can you have a look to it and remove your blocking?

**Review by ddiezrod — APPROVED** [review]
Solution seems logic to me, tests from our side are running fine. 👍

**Comment by ddiezrod** [comment]
@philbucher I am not sure if I understand your point. At the end the nodes only have one variable list, so you can't have two model parts with different lists. If you want that, then you need to have two completely different model parts duplicating nodes and elements (I dont know if there is already something like this) and then managing the communication somehow on your own way. I think that is not the behaviour we want to have for this utility.

**Comment by philbucher** [comment]
No worries, please go ahead :)

**Comment by ddiezrod** [comment]
Ok! Can we merge then @rubenzorrilla ?

**Comment by rubenzorrilla** [comment]
> Ok! Can we merge then @rubenzorrilla ?

Sure. I was waiting for @pooyan-dadvand 's opinion but I don't think there will be any problem.

**Comment by pooyan-dadvand** [comment]
Just for the record,

When the variables list where belong to modelpart , the problem was that we could delete the model part with the  variables list and the node which was shared with other model part would leave pointing to an empty variables list. So modelpart would clean the nodes container to make it more clear that the nodal data is not accessible and avoiding memory leak. 

Now with variables lists having intrusive pointer we don't need this anymore and we can delete the nodal data in its destructor as it should be.
